handlers/blocker.go

```diff
@@ -0,0 +1,427 @@
+			return
+		}
+
+		b.log.Debug("Connection returned to pool (from Blocker)")
+
+		if err := b.parent.delete("b", b.source); err != nil {
+			b.log.Warn("Could not remove blocker from reservation list", zap.Error(err))
+		}
+
+		for i := range b.queue {
+			local := b.queue[i].local
+			local.Close()
+
+			b.parent.monitor.decrementLocalBlockers(
+				"reservation_event.local_unblock",
+				[]string{fmt.Sprintf("source:%s", b.source)})
+		}
+
+		b.parent.monitor.decrementUpstreamBlockers(
+			"reservation_event.upstream_unblock",
+			[]string{fmt.Sprintf("source:%s", b.source)})
+	})
+}
+
+func (b *blocker) roundTrip(cmd *blockingCommand) error {
+	if err := b.upstreamWrite(cmd.command); err != nil {
+		return fmt.Errorf("upstream write: %s", err)
+	}
+
+	res, err := b.upstreamRead(b.readTimeout)
+	if err != nil {
+		if errors.Is(err, io.EOF) {
+			return upstreamDisconnectedErr
+		}
+
+		return fmt.Errorf("upstream read: %w", err)
+	}
+
+	if isClusterResponse(res) {
+		err = fmt.Errorf("blocking commands aren't supported for Redis cluster")
+		mm := []*redis.Message{redis.NewError([]byte(fmt.Sprintf("redisbetween: %v", err.Error())))}
+		b.localWrite(cmd.local, mm)
+
+		return err
+	}
+
+	if cmd.local.Closed() {
+		return fmt.Errorf("local connection was closed")
+	}
+
+	if err := b.localWrite(cmd.local, res); err != nil {
+		return fmt.Errorf("local write: %w", err)
+	}
+
+	return nil
+}
+
+// Convenience wrapper for WriteWireMessage for reservation upstreams
+func (b *blocker) upstreamWrite(wm []*redis.Message) error {
+	err := b.messenger.Write(
+		b.ctx, b.log, wm, b.upstream.Conn(), b.upstream.Address().String(),
+		b.upstream.ID(), b.writeTimeout, false, b.upstream.Close,
+	)
+
+	return err
+}
+
+// Convenience wrapper for ReadWireMessage for reservation upstreams
+func (b *blocker) upstreamRead(timeout time.Duration) ([]*redis.Message, error) {
+	msg, err := b.messenger.Read(
+		b.ctx, b.log, b.upstream.Conn(), b.upstream.Address().String(),
+		b.upstream.ID(), b.readTimeout, 1, false, b.upstream.Close,
+	)
+
+	return msg, err
+}
+
+func (b *blocker) localWrite(c *connection, wm []*redis.Message) error {
+	return c.Write(wm)
+}
+
+func parseCommand(c *connection, wm []*redis.Message) (*blockingCommand, error) {
+	var parts [][]byte
+
+	for _, part := range wm[0].Array {
+		parts = append(parts, part.Value)
+	}
+
+	if len(parts) < 4 {
+		return nil, fmt.Errorf("invalid command: %v", wm)
+	}
+
+	timeout, err := time.ParseDuration(fmt.Sprintf("%ss", parts[3]))
+	if err != nil {
+		return nil, err
+	}
+
+	cmd := &blockingCommand{
+		local:   c,
+		command: wm,
+		source:  string(parts[1]),
+		timeout: timeout,
+		start:   time.Now(),
+	}
+
+	return cmd, nil
+}
+
+// Checks if command is a) singular and b) is a *blocking command
+func isBlockingCommand(commands []string) bool {
+	if len(commands) == 1 && strings.Contains(commands[0], "BRPOPLPUSH") {
+		return true
+	}
+
+	return false
+}
```

Comment:
I think this should also include

```
BLMOVE
BLMPOP
BLPOP
BRPOP
```

Reply:
At this point, this is just adding `BRPOPLPUSH` but others wouldn’t be too difficult to add eventually.